src/Namespace.js

```diff
@@ -0,0 +1,73 @@
+// @flow
+
+import ParseError from "./ParseError";
+
+export type Mapping<Value> = {[string]: Value};
+
+export default class Namespace<Value> {
+    current: Mapping<Value>;
+    builtins: Mapping<Value>;
+    undefStack: Mapping<Value>[];
+
+    /**
+     * Both arguments are optional.  The first argument is an object of built-in
+     * mappings which never change.  The second argument is an object of initial
+     * mappings, which will constantly change according to any `set`s done.
+     */
+    constructor(builtins: Mapping<Value> = {}, topGroup: Mapping<Value> = {}) {
+        this.current = topGroup;
+        this.builtins = builtins;
+        this.undefStack = [];
+    }
+
+    beginGroup() {
+        this.undefStack.push({});
+    }
+
+    endGroup() {
+        if (this.undefStack.length === 0) {
+            throw new ParseError("Unbalanced namespace destruction: attempt " +
+                "to pop global namespace; please report this as a bug");
+        }
+        const undefs = this.undefStack.pop();
+        for (const undef of Object.getOwnPropertyNames(undefs)) {
+            if (undefs[undef] === undefined) {
+                delete this.current[undef];
+            } else {
+                this.current[undef] = undefs[undef];
+            }
+        }
+    }
+
+    get(name: string): ?Value {
+        if (this.current.hasOwnProperty(name)) {
+            return this.current[name];
+        } else {
+            return this.builtins[name];
```

Comment:
No, it can't.  (I assume you mean `||`, as `|` doesn't short circuit.  I used to have it this way, but it fails tests.)  We need to distinguish a macro set to the empty string (which evaluates to `false`) vs. an `undefined` value.